url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/tffst/Tffst.java

```diff
@@ -0,0 +1,1712 @@
+   * 
+   * @see #setInitialState(State)
+   * @see State
+   * @see Transition
+   */
+  public Tffst() {
+    this.initial = new State();
+    this.deterministic = true;
+  }
+
+  /**
+   * Assign another Tffst to this
+   * 
+   * @param Tffst
+   */
+  public void assign(Tffst tffsr) {
+    this.initial = tffsr.initial;
+    this.deterministic = tffsr.deterministic;
+    this.info = tffsr.info;
+  }
+
+  /**
+   * Returns a clone of this Tffst.
+   * @throws CloneNotSupportedException 
+   */
+  public Tffst clone() throws CloneNotSupportedException {
+
+    Tffst a = new Tffst();
+    HashMap<State, State> m = new HashMap<State, State>();
+    Set<State> states = getStates();
+
+    for ( State s : states) {
+      State p = m.get(s);
+      if (p == null) {
+        p = new State();
+        m.put(s, p );
+      }
+      p.accept = s.accept;
+      if (s == initial) a.initial = p;
+    }
+    for ( State s : states) {
+      State p = m.get(s);
+      for (Transition t : s.getTransitions()){
+        new Transition(p, t.getLabelIn().clone(), t.getLabelOut().clone(),m.get(t.getTo()));
+      }
+    }
+    a.deterministic = deterministic;
+    a.info = info;
+    return a;
+  }
+
+  /**
+   * COMPOSITION
+   * this o theOther (i.e. theOther(this) )   
+   * 
+   * Constraint: this cannot have e-input labels and theOther cannot have e-output labels.  
+   * 
+   * ct((p1,π1,π1,q1,1),(p2,π2,π2,q2,1))=((p1,p2),π1∧π2,π1∧π2,(q1,q2),1)
+   * ct((p1,φ,π1,q1,0),(p2,π2,π2,q2,1))=((p1,p2),φ,π1∧π2,(q1,q2),0)
+   * ct((p1,π1,π1,q1,1),(p2,π2,ψ,q2,0))=((p1,p2),π1∧π2,ψ,(q1,q2),0)
+   * ct((p1,φ,π1,q1,0),(p2,π2,ψ,q2,0))=((p1,p2),φ,ψ,(q1,q2),0) if
+   * satisfiable(π1∧π2)
+   * 
+   * E= {ct(e1,e2 )|e1 ∈ E1 , e2 ∈ E2 } ∪ {((p1,p2 ), ,ψ,(p1,q2),0)|p1 ∈
+   * Q1,(p2,,ψ,q2,0) ∈ E2 } ∪ {((p1,p2 ),φ, ,(q1,p2),0)|p2 ∈ Q2,(p1,φ,,q1,0) ∈
+   * E1 }
+   * 
+   * 
+   * 
+   * 
+   */
+  
+  public Tffst composition(Tffst theOther) { 
+
+    Tffst thisSimple = this.toSingleLabelTransitions();
+    Tffst theOtherSimple = theOther.toSingleLabelTransitions();
+    
+    Tffst out = new Tffst();
+
+    Set<State> thisStates = thisSimple.getLiveStates();
+    Set<State> theOtherStates = theOtherSimple.getLiveStates();
+
+    HashMap<StatePair, State> newstates = new HashMap<StatePair, State>();
+
+    for (State thisState : thisStates){
+      for (State theOtherState : theOtherStates){
+        
+        // get the new state that is equivalent to this pair, create it. 
+        State outStartState = newstates.get(new StatePair(thisState, theOtherState));
+        if (outStartState == null) {
+          outStartState = new State();
+          newstates.put(new StatePair(thisState, theOtherState), outStartState);
+          if (thisState.equals(thisSimple.initial) && theOtherState.equals(theOtherSimple.initial))
+            out.initial = outStartState;
+          outStartState.setAccept(thisState.isAccept() && theOtherState.isAccept());
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 351:357, 457:463.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.